File: 21. Merge Two Sorted Lists.cc
//Using Recursive method
//TC: O(n + m), where n is the number of nodes in list1 and m is the number of nodes in list2.
//SC: O(n + m), where n + m call times are stored on recursive call stack

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (list1 == nullptr) {
            return list2;   
        }
        if (list2 == nullptr) {
            return list1;
        }
        
        if (list1->val <= list2->val) {
            list1->next = mergeTwoLists(list1->next, list2);
            return list1;
        } else {
            list2->next = mergeTwoLists(list1, list2->next);
            return list2;
        }
        
    }

};

File: java/21.Merge_Two_Sorted_Lists.java
// TC: O(m + n)
// SC: O(1)
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null || list2 == null)
            return list1 != null ? list1 : list2;
        
        ListNode res = new ListNode();
        ListNode it = res;
        while (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                it.next = list1;
                list1 = list1.next;
            } else {
                it.next = list2;
                list2 = list2.next;
            }
            it = it.next;
        }
        it.next = list1 != null ? list1 : list2;
        return res.next;
    }
}
